File: frontend/lib/planner/models/page_response.dart
/// Stable pagination envelope, mirroring {@code PageResponse}.
class PageResponse<T> {
  PageResponse({
    required this.content,
    required this.page,
    required this.size,
    required this.totalElements,
    required this.totalPages,
    required this.last,
  });

  final List<T> content;
  final int page;
  final int size;
  final int totalElements;
  final int totalPages;
  final bool last;

  factory PageResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Map<String, dynamic>) itemFromJson,
  ) {
    final rawContent = json['content'];
    return PageResponse<T>(
      content: [
        if (rawContent is List)
          for (final item in rawContent)
            if (item is Map<String, dynamic>) itemFromJson(item),
      ],
      page: (json['page'] as num?)?.toInt() ?? 0,
      size: (json['size'] as num?)?.toInt() ?? 0,
      totalElements: (json['totalElements'] as num?)?.toInt() ?? 0,
      totalPages: (json['totalPages'] as num?)?.toInt() ?? 0,
      last: json['last'] == true,
    );
  }
}

File: backend/src/main/java/com/blistra/planner/dto/PageResponse.java
package com.blistra.planner.dto;

import java.util.List;

/**
 * A stable page envelope returned by collection endpoints. Keeps the pagination
 * contract explicit and independent of Spring's internal Page type.
 */
public record PageResponse<T>(List<T> content, int page, int size, long totalElements, int totalPages, boolean last) {

    public static <T> PageResponse<T> of(org.springframework.data.domain.Page<T> page) {
        return new PageResponse<>(
                page.getContent(),
                page.getNumber(),
                page.getSize(),
                page.getTotalElements(),
                page.getTotalPages(),
                page.isLast());
    }
}